### PR TITLE
Enable source build for ASP.NET external access project

### PR DESCRIPTION
**Pull request by JamesNK**
~Also needs to be applied to the 17.3 branch.~ Roslyn appears to merge lower branches into higher, so this should go straight into 17.3 once approved.

**Change:**
Enable source build for `Microsoft.CodeAnalysis.ExternalAccess.AspNetCore`. This change requires removing some dependencies and functionality that don't support source build. They will be worked around on ASP.NET Core side.

**Reason for change:**
ASP.NET Core references `Microsoft.CodeAnalysis.ExternalAccess.AspNetCore` from an analyzer project. Because of aspnetcore's source build requirement, source build needs to be enabled for this project. The feature is targeting .NET 7 lines up with Microsoft.CodeAnalysis 4.3.x assemblies and dev17.3.

**Risk**:
Low. `Microsoft.CodeAnalysis.ExternalAccess.AspNetCore` is only referenced by ASP.NET Core code.

cc @CyrusNajmabadi

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by JamesNK** [comment]
Changes fail in the source build job:
https://dev.azure.com/dnceng/public/_build/results?buildId=1905314&view=logs&j=c316b76e-5b95-5e79-2740-62e92f13eebd&t=79bcefda-4e9d-5f89-72b6-b7cfe8a925eb&l=172

>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(22,61): error CS0246: The type or namespace name 'IEmbeddedLanguageBraceMatcher' could not be found (are you missing a using directive or an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(22,61): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'IEmbeddedLanguageBraceMatcher' could not be found (are you missing a using directive or an assembly reference?)
>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(8,30): error CS0234: The type or namespace name 'BraceMatching' does not exist in the namespace 'Microsoft.CodeAnalysis' (are you missing an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(8,30): error CS0234: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'BraceMatching' does not exist in the namespace 'Microsoft.CodeAnalysis' (are you missing an assembly reference?)
>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/EmbeddedLanguages/BraceMatching/AspNetCoreBraceMatchingResult.cs(15,18): error CS0246: The type or namespace name 'BraceMatchingResult' could not be found (are you missing a using directive or an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/EmbeddedLanguages/BraceMatching/AspNetCoreBraceMatchingResult.cs(15,18): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'BraceMatchingResult' could not be found (are you missing a using directive or an assembly reference?)
>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(16,6): error CS0246: The type or namespace name 'ExportEmbeddedLanguageBraceMatcherAttribute' could not be found (are you missing a using directive or an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(16,6): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'ExportEmbeddedLanguageBraceMatcherAttribute' could not be found (are you missing a using directive or an assembly reference?)
>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(16,6): error CS0246: The type or namespace name 'ExportEmbeddedLanguageBraceMatcher' could not be found (are you missing a using directive or an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(16,6): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'ExportEmbeddedLanguageBraceMatcher' could not be found (are you missing a using directive or an assembly reference?)
>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(35,13): error CS0246: The type or namespace name 'BraceMatchingOptions' could not be found (are you missing a using directive or an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(35,13): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'BraceMatchingOptions' could not be found (are you missing a using directive or an assembly reference?)
>   /__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(30,16): error CS0246: The type or namespace name 'BraceMatchingResult' could not be found (are you missing a using directive or an assembly reference?) [/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj]
> ##[error]artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore/Internal/EmbeddedLanguages/AspNetCoreEmbeddedLanguageBraceMatcher.cs(30,16): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'BraceMatchingResult' could not be found (are you missing a using directive or an assembly reference?)

Is the InternalsVisibleTo pattern a problem here?

**Comment by tmat** [comment]
I don't think we can build EditorFeatures from source.

**Comment by JamesNK** [comment]
I think only brace matching uses `Microsoft.CodeAnalysis.EditorFeatures`. A solution could be to remove the brace matching API and that dependency from this project.

**Comment by JamesNK** [comment]
@MichaelSimons Is source build enabled for this project once this PR is merged? Is anything else required to register it and unblock the ASP.NET Core route tooling changes?

**Comment by MichaelSimons** [comment]
> @MichaelSimons Is source build enabled for this project once this PR is merged? Is anything else required to register it and unblock the ASP.NET Core route tooling changes?

Source-build is enabled for this repo.  To validate, retrieve the `*SourceBuild.Intermediate*` package created by the `Source-build (Managed)` CI leg and check that it includes the Microsoft.CodeAnalysis.ExternalAccess.AspNetCore package.

**Comment by JamesNK** [comment]
@CyrusNajmabadi I've switched PR to main. Can you take a look?

File: src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj
Comment (JamesNK) [inline src/Tools/ExternalAccess/AspNetCore/Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.csproj:15]:
```suggestion
    <!-- Referenced by dotnet/aspnetcore and included in SDK. Requires source build -->
```

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
> Source-build is enabled for this repo. To validate, retrieve the *SourceBuild.Intermediate* package created by the Source-build (Managed) CI leg and check that it includes the Microsoft.CodeAnalysis.ExternalAccess.AspNetCore package.

@JamesNK, before I merge, did you validate that the result package has what you need?

**Comment by JamesNK** [comment]
I couldn't find the SourceBuild.Intermediate package in artifacts, but I can see it's being built in logs: `/__w/1/s/artifacts/source-build/self/src/src/Tools/ExternalAccess/AspNetCore`

**Pull request merged**
